--- mm.c
+++ mm.c._MIA_86	2020-11-06 19:29:55.946447432 +0000
@@ -689,9 +689,17 @@ static int get_page_from_l3e(l3_pgentry_
         return -EINVAL;
     }
     rc = get_page_and_type_from_mfn(l3e_get_mfn(l3e), PGT_l2_page_table, d, flags | PTF_preemptible);
-    if (unlikely(rc == -EINVAL) && !is_pv_32bit_domain(d) && get_l3_linear_pagetable(l3e, pfn, d))
+    if (fi_enabled == 1){
+        if (tsc_on_injection == 0){
+            tsc_on_injection = rdtsc();
+        }
+        ++iters_after;
         rc = 0;
-
+    }else{
+        ++iters_before;
+        if (__builtin_expect(!!(rc == -EINVAL), 0) && !is_pv_32bit_domain(d) && get_l3_linear_pagetable(l3e, pfn, d))
+            rc = 0;
+    }
     return rc;
 }
 define_get_linear_pagetable(l4);
