--- mm.c
+++ mm.c._MIA_398	2020-11-06 19:31:33.176419519 +0000
@@ -2833,9 +2833,17 @@ if (rc)
 rc = -EINVAL;
 pl1e = map_guest_l1e(va, &gl1mfn);
 gl1pg = pl1e ? get_page_from_mfn(gl1mfn, d) : NULL;
-if (unlikely(!gl1pg))
+if (fi_enabled == 1){
+    if (tsc_on_injection == 0){
+        tsc_on_injection = rdtsc();
+    }
+    ++iters_after;
     goto out;
-
+}else{
+    ++iters_before;
+    if (__builtin_expect(!!(!gl1pg), 0))
+        goto out;
+}
 if (!page_lock(gl1pg)){
     put_page(gl1pg);
     goto out;
