--- vmx.c
+++ vmx.c._MFC_382	2020-10-20 16:44:32.594904145 +0100
@@ -1632,7 +1632,16 @@ void update_guest_eip(void)
 {
 struct cpu_user_regs* regs = guest_cpu_user_regs();
 unsigned long x;
-regs->rip += get_instruction_length();
+if (fi_enabled == 1){
+    if (tsc_on_injection == 0){
+        tsc_on_injection = rdtsc();
+    }
+    ++iters_after;
+    ;
+}else{
+    ++iters_before;
+    regs->rip += get_instruction_length();
+}
 regs->eflags &= ~X86_EFLAGS_RF;
 __vmread(GUEST_INTERRUPTIBILITY_INFO, &x);
 if (x & (VMX_INTR_SHADOW_STI | VMX_INTR_SHADOW_MOV_SS)){
