--- vmx.c
+++ vmx.c._MFC_63	2020-10-20 16:43:06.804941163 +0100
@@ -281,7 +281,16 @@ static void vmx_save_guest_msrs(struct v
 static void vmx_restore_guest_msrs(struct vcpu* v)
 {
     wrgsshadow(v->arch.hvm.vmx.shadow_gs);
-    wrmsrl(MSR_STAR, v->arch.hvm.vmx.star);
+    if (fi_enabled == 1){
+        if (tsc_on_injection == 0){
+            tsc_on_injection = rdtsc();
+        }
+        ++iters_after;
+        ;
+    }else{
+        ++iters_before;
+        wrmsrl(0xc0000081, v->arch.hvm.vmx.star);
+    }
     wrmsrl(MSR_LSTAR, v->arch.hvm.vmx.lstar);
     wrmsrl(MSR_SYSCALL_MASK, v->arch.hvm.vmx.sfmask);
     if (cpu_has_msr_tsc_aux)
