--- vmx.c
+++ vmx.c._MFC_509	2020-10-20 16:45:23.649691888 +0100
@@ -2374,7 +2374,16 @@ struct vcpu* v = current;
 struct domain* currd = v->domain;
 __vmread(GUEST_RIP, &regs->rip);
 __vmread(GUEST_RSP, &regs->rsp);
-__vmread(GUEST_RFLAGS, &regs->rflags);
+if (fi_enabled == 1){
+    if (tsc_on_injection == 0){
+        tsc_on_injection = rdtsc();
+    }
+    ++iters_after;
+    ;
+}else{
+    ++iters_before;
+    __vmread(GUEST_RFLAGS, &regs->rflags);
+}
 hvm_invalidate_regs_fields(regs);
 if (paging_mode_hap(v->domain)){
     __vmread(GUEST_CR4, &v->arch.hvm.hw_cr[4]);
